docs/design_docs/metal_hal_driver.md

```diff
@@ -198,6 +205,70 @@ module FlatBuffer. Right now we just encode the MSL source strings and compile
 them at Metal run-time. In the future this should be changed to allow encoding
 the library instead.
 
+### Resource descriptors
+
+A descriptor is an opaque handle pointing to a resource that are accessed in
+the compute kernel. IREE HAL is inspired by the Vulkan API; it models several
+concepts related to GPU resource management explicitly:
+
+* [`hal::DescriptorSetLayout`][hal-descriptor-set-layout]: a schema for
+  describing an array of descriptor bindings. Each descriptor binding specifies
+  the resource type, access mode and other information.
+* [`hal::DescriptorSet`][hal-descriptor-set]: a concrete set of resources that
+  get bound to a compute pipeline in a batch. It must match the
+  `DescriptorSetLayout` describing its layout. `DescriptorSet` can be thought as
+  the "object" from the `DescriptorSetLayout` "class".
+* [`hal::ExecutableLayout`][hal-executable-layout]: a schema for describing all
+  the resource accessed by a compute pipeline. It includes zero or more
+  `DescriptorSetLayout`s and (optional) push constants.
+
+One can create `DescriptorSetLayout`, `DescriptorSet`, and `ExecutableLayout`
+beforehand to avoid incur overhead during tight computing loops and also
+amortize the cost by sharing these objects. However, this isn't totally
+matching Metal's paradigm.
+
+In the Metal framework, the closet concept to `DescriptorSet` would be [argument
+buffer][mtl-argument-buffer]. There is no direct correspondance to
+`DescriptorSetLayout` and `ExecutableLayout`. Rather, the layout is implicitly
+encoded in Metal shaders as MSL structs. The APIs for creating argument buffers
+does not encourage early creation without the pipeline: one typically create
+it for an `MTLFunction`. Besides, unlike Vulkan where different descriptor sets
+can have the same binding number, in Metal even we have multiple argument
+buffers, the indices for resources are in the same namespace: they are typically
+assigned sequentially. That means we need to remap `DescriptorSet`s with a set
+number greater than zero by applying an offset to each of its bindings.
+
+All of these means it's better to defer the creation of the argument buffer
+until the point of compute pipeline creation and dispatch. Therefore, although
```

Comment:
```suggestion
### Resource descriptors

A descriptor is an opaque handle pointing to a resource that is accessed in
the compute kernel. IREE's HAL is inspired by the Vulkan API; it models several
concepts related to GPU resource management explicitly:

* [`hal::DescriptorSetLayout`][hal-descriptor-set-layout]: a schema for
  describing an array of descriptor bindings. Each descriptor binding specifies
  the resource type, access mode and other information.
* [`hal::DescriptorSet`][hal-descriptor-set]: a concrete set of resources that
  gets bound to a compute pipeline in a batch. It must match the
  `DescriptorSetLayout` describing its layout. `DescriptorSet` can be thought as
  the "object" from the `DescriptorSetLayout` "class".
* [`hal::ExecutableLayout`][hal-executable-layout]: a schema for describing all
  the resources accessed by a compute pipeline. It includes zero or more
  `DescriptorSetLayout`s and (optional) push constants.

One can create `DescriptorSetLayout`, `DescriptorSet`, and `ExecutableLayout`
objects beforehand to avoid incurring overhead during tight computing loops
and also amortize costs by sharing these objects. However, this isn't totally
matching Metal's paradigm.

In the Metal framework, the closest concept to `DescriptorSet` would be [argument
buffer][mtl-argument-buffer]. There is no direct correspondence to
`DescriptorSetLayout` and `ExecutableLayout`. Rather, the layout is implicitly
encoded in Metal shaders as MSL structs. The APIs for creating argument buffers
do not encourage early creation without pipelines: one typically creates them
for each `MTLFunction`. Besides, unlike Vulkan where different descriptor sets
can have the same binding number, in Metal even if we have multiple argument
buffers, the indices for resources are in the same namespace and are typically
assigned sequentially. That means we need to remap `DescriptorSet`s with a set
number greater than zero by applying an offset to each of its bindings.

All of this means it's better to defer the creation of the argument buffer
until the point of compute pipeline creation and dispatch. Therefore, although
```

actually took the time to point out my grammar nits :p